components/core/src/clp/BoundedReader.hpp

```diff
@@ -0,0 +1,78 @@
+#ifndef CLP_BOUNDEDREADER_HPP
+#define CLP_BOUNDEDREADER_HPP
+
+#include "ReaderInterface.hpp"
+
+namespace clp {
+/**
+ * BoundedReader is a ReaderInterface designed to wrap other ReaderInterfaces and prevent users
+ * from reading or seeking beyond a certain point in the underlying input stream.
+ *
+ * This is useful when the underlying input stream is divided into several logical segments and we
+ * want to prevent a reader for an earlier segment consuming any bytes from a later segment. In
+ * particular, reading part of a later segment may force the reader for that later segment to seek
+ * backwards, which can be either inefficient or impossible for certain kinds of input streams.
+ */
+class BoundedReader : public ReaderInterface {
+public:
+    // Constructor
+    explicit BoundedReader(ReaderInterface* reader, size_t bound)
+            : m_reader(reader),
+              m_bound(bound) {
+        if (nullptr == m_reader) {
+            throw ReaderInterface::OperationFailed(ErrorCode_BadParam, __FILE__, __LINE__);
+        }
+        m_curr_pos = m_reader->get_pos();
+        if (m_curr_pos > m_bound) {
+            throw ReaderInterface::OperationFailed(ErrorCode_BadParam, __FILE__, __LINE__);
+        }
+    }
+
+    // Methods implementing the ReaderInterface
+    /**
+     * Tries to get the current position of the read head in the underlying reader.
+     * @param pos Position of the read head in the underlying reader
+     * @return ErrorCode_Success on success
+     * @return ErrorCode_errno on failure
+     */
+    auto try_get_pos(size_t& pos) -> ErrorCode override { return m_reader->try_get_pos(pos); }
+
+    /**
+     * Tries to seek to the given position, limited by the checkpoint. If pos is past the checkpoint
+     * then this reader seeks up to the checkpoint in the underlying stream, then returns
+     * ErrorCode_EndOfFile on success or ErrorCode_errno otherwise.
```

Comment:
This description is duplicates the return section. I think we can drop the failure case.
```suggestion
     * Tries to seek to the given position, limited by the checkpoint.
```